GIP: Optimistic EBO (#53)
graphprotocol/graph-improvement-proposals: gips/0080-optimistic-ebo.md

---
GIP: "0080"
Title: Optimistic EBO

> Authors: mono <0xmono@defi.sucks>, shaito <shaito@defi.sucks>

Reviewer: ```suggestion
Authors: "mono <0xmono@defi.sucks>", "shaito <shaito@defi.sucks>"
```

# **Prior Art**

Several approaches have been proposed to improve the oracle process within The Graph protocol:

- [Submitting recent blocks for POIs](https://forum.thegraph.com/t/require-that-more-recent-pois-are-submitted-in-order-to-collect-indexing-rewards-multi-blockhain-pois/2500)
- [On-chain epoch oracle](https://hackmd.io/KHLuc4haR5uCnmpMu0Ce7Q)

> Before adopting the optimistic approach, expanding the number of EBO operators to 3-5 Council-authorized organizations was considered. However, the design proposed in this GIP provides greater decentralization, reducing the potential for coordinated corruption and the reliance on the Council as a single point of failure.
> 

## Current status of Epoch Block Oracle

The EBO defines the source of truth for the block number corresponding to a new epoch in each indexed chain. It fetches the block for each epoch on each chain and records it on-chain on the protocol chain (Arbitrum One). An epoch is a period determined by the [EpochManager](https://github.com/graphprotocol/contracts/blob/main/contracts/epochs/EpochManager.sol), which lasts approximately one day but is measured in blocks.

> Indexers use the EBO subgraph to get the block number for each chain and use it to build the POI, which is published when closing allocations to claim rewards.

Reviewer: While most people know what this is, I suggest at least once in the document being explicit.

```suggestion
Indexers use the EBO subgraph to get the block number for each chain and use it to build the POI (Proof Of Indexing), which is published when closing allocations to claim rewards.
```

If a dispute arises, an indexer who publishes a POI for the wrong block may be disputed by a fisherman and penalized by the Arbitrator.

The EBO is composed of 3 parts:

- A Rust process that reads information from RPCs.
- A [Data Edge](https://github.com/graphprotocol/block-oracle/blob/main/packages/contracts/contracts/DataEdge.sol) contract where updates get posted (compressed and encoded).

> - A [Subgraph](https://github.com/graphprotocol/block-oracle/tree/main/packages/subgraph) that indexes the Data Edge allows indexers to craft the POI easily.

Reviewer: ```suggestion
- A [Subgraph](https://github.com/graphprotocol/block-oracle/tree/main/packages/subgraph) that indexes the Data Edge allowing indexers to easily craft their POI.
```

Key improvements:

- **Optimistic operators:** Implementing an optimistic oracle will allow EBO operations to be open and permissionless, improving decentralization, redundancy, and scalability.
- **Timestamp comparison:** Instead of querying each RPC when detecting a new epoch, a delay is introduced to compare timestamps, reducing time sensitivity and improving the precision of fork selection.
- **Escalating bond mechanism**: The modularity of the optimistic oracle allows for an escalating bond dispute system, reducing Arbitrator intervention and enabling lower initial stakes without compromising security.
    - The Arbitrator multisig will act as the ultimate source of truth in the case of disputes in this initial phase, with potential future improvements.

### Optimistic Design

![image.png](../assets//gip-0080/Optimistic%20EBO.png)

### **Optimistic operators**

The optimistic approach allows off-chain agents to propose updates to the Optimistic EBO, which other participants can dispute by staking tokens or using alternative resolution mechanisms to align incentives and ensure accuracy.

> For this, we will use [Prophet](https://github.com/defi-wonderland/prophet-modules/tree/dev), an optimistic, modular, and public-good oracle that abstracts the complexity of the proposal and dispute process while offering the flexibility to customize its functionality as needed. For more details on Prophet, refer to the documentation [here](https://docs.prophet.tech/content/modules/index.html).

Reviewer: https://heyoka.medium.com/dont-use-click-here-f32f445d1021
While at it, cleaner URL.
```suggestion
For this, we will use [Prophet](https://github.com/defi-wonderland/prophet-modules/tree/dev), an optimistic, modular, and public-good oracle that abstracts the complexity of the proposal and dispute process while offering the flexibility to customize its functionality as needed. For more details see [Prophet documentation](https://docs.prophet.tech/content/modules/).
```

> ~~~~

Reviewer: Is this meant to be an HR? It seems to be causing subsequent text to render as code. Might need a blank line or another change.

The process would work as follows: 

- Any staked actor can propose an update for the EBO, which is published in a contract and remains open for a set period to allow for potential disputes.
- Any staked actor can also challenge the proposed update. If a dispute arises, the initial proposal is discarded, and a new one can be submitted.

> - If a dispute occurs, the resolution mechanism is triggered, and the winner of the dispute will receive the counterparty’s bond as a reward.

Reviewer: To avoid certain attack vectors, this should be not the full bond, but half of it?

Author: Which attack vectors? What do we do with the other half?

Reviewer: If a disputed fault does not rely in a loss, then that potentially creates attack vectors of the malicious actor being both the proposer and the disputer. Depending on protocol details and circumstances this might allow providing bad data without cost or process delays.

- If no disputes are raised within the designated time window, or if the dispute is resolved in favor of the proposer, the update is considered valid.

Operators can update the block number of one chain at a time, allowing for ongoing disputes and resolutions for each updated chain. This improves the system's scalability.

> The [Detailed Specification](https://www.notion.so/DRAFT_I-GIP-0XYZ-Optimistic-EBO-74a0b9df7008464e9dc1215c46920213?pvs=21) section will provide further details on how contracts and off-chain agents interact.

Reviewer: That specification is not, and likely will not be, publically visible? We should only link to information in the public domain?

The following methods can be used to prove that a specific block number corresponds to a claimed timestamp:

- Check the timestamps of two consecutive blocks $t_N$ and $t_{N+1}$. If the timestamp value from the protocol chain $t_0$ lives in the range $[t_N,t_{N+1}]$, then $N$ is the first block corresponding to the new epoch.
- Check a single block timestamp $t_N$ and an average block measure defined per chain $\Delta_i$. If the timestamp value from the protocol chain $t_0$ lives in the range $[t_N,t_N+\Delta_i]$, then $N$ is the first block corresponding to the new epoch.

### Escalating bond

To reduce the number of calls to The Graph's Arbitrator, we will implement an escalating bond system:

- Anyone can dispute while the dispute window is open and if staked. To dispute, you need to post a bond.
    - The resolution system is triggered once the participants have completed the maximum number of rounds.
    - If the disputant “overtakes” the Min Bond, the answer is considered incorrect until the proposer responds. The proposer must overtake the disputant's Bond to escalate the answer further. This process can be repeated up to N times.

> - After each dispute, the original update gets deleted and can be answered again, even before resolution. This option makes the process quicker for the requestor. This answer can be disputed using the same method as described above.

Reviewer: Likely want to use one consistent spelling in the document.
```suggestion
- After each dispute, the original update gets deleted and can be answered again, even before resolution. This option makes the process quicker for the requester. This answer can be disputed using the same method as described above.
```

    - An honest proposer will be rewarded more for the won resolution than the original update.
    - Disputers are incentivized to post the new answer, as they have already computed the query.

> - The payment from the requester will end up going to the proposer with the correct answer

Reviewer: Term 'requester' is used just twice (with different spellings) and not defined. Is the requester the dispute raiser (might be simpler to say disputer)?

I believe that this should not be the entire payment.

Reviewer 2: The requester is the one who requests data from Prophet and also sets the rewards for proposers to submit responses. The disputer monitors the proposals and challenges them if they are invalid.

The “overtake” mechanism improves resolution time and reduces Arbitrator calls, while the second answer (before resolution) improves the answering time to the request. 

> ![EBO Designs (6).png](../assets/gip-0080/Optimistic%20EBO%20Designs%20(1).png)

Reviewer: I wonder if better to cleanup image filenames to not have version suffixes such as "(1)", and not have spaces in the filename?

> Using an Arbitrator (a Council of qualified individuals selected by the governance) is the simplest way to launch the system at this early stage and manage all the chains with a single resolution mechanism. This approach allows for greater decentralization in the future, as the source of truth can be easily migrated.
> 

**Some considerations:**

- Implementing multiple resolution systems by chain or type (rollups, L1s, etc) adds complexity.
- The Council should upgrade the system to accommodate future improvements.
- Disputes should be managed per-chain to prevent a single dispute from impacting the entire system.

## **Economic Security considerations**

We recognize that economic security in The Graph presents complex challenges, which various teams are researching. The ideas discussed here are preliminary and may adapt to future protocol developments.

### Parameters

We suggest the following parameter values for the Escalation Bond system:

- **Bond threshold**: The amount of stake that triggers a call to the Arbitrator (virtual bond in the escalation system), set at 100K GRT.
- **Round**: The maximum number of rounds a dispute can escalate, set at 5.

> - **Min Bond**: Minimum amount required for five rounds of escalations to reach the Bond threshold, set at 11111 GRT.

Reviewer 2: It's a little confusing and I'm rewriting this section.

MinBond is the amount you need to initiate a dispute. Then the proposers must overtake the minbond by doubling the bond until the proposer or disputer stops answering or the total amount bet exceeds the bond threshold and calls the arbitrator. In this case it would look like this:

Round 0 - (start dispute): 11111
Round 1- proposer 22222
Round 2- disputer 22222
Round 3- proposer 22222
Round 4- disputer 22222
Round 5- proposer 22222 -> call arbitrator

> - **Min Bond**: Minimum amount required for five rounds of escalations to reach the Bond threshold, set at 11111 GRT.

Author: @RembrandtK The bond gets increased by 1 bond unit (configured in the request) for every escalation. So if we set 11111 GRT as the bond, every escalation increases it by that amount

- **Dispute Window**: An indexer's maximum response time to a dispute is 30 minutes.

> These values are based on an economic security analysis that evaluates the potential gains from corruption across various attack scenarios. You can read the complete analysis [here](https://www.notion.so/Economic-security-considerations-1009a4c092c780b6afb1fa8bb8c266be?pvs=21).

Reviewer: Not publicy accessible. Also note: https://heyoka.medium.com/dont-use-click-here-f32f445d1021

**The Arbitrator has the authority to modify these parameters.**

# **Detailed Specification**

## Prophet modules

Interaction with Prophet requires the following modules that we will use/build to support the EBO:

> - **RequestModule:** Used to request data in prophet. We will build the EBORequestModule which will allow users to create epoch-block requests to the oracle with the needed data for it. Compatible with partial requests (a subset of all the chains) and will contain the epoch number and the chain to fetch for that epoch.

Reviewer: I note sometimes ":" is bold too, other times not. I prefer not personally, but mainly suggest being consistent.

- **ResponseModule:** Used by data providers to answer requests.

> - **DisputeModule:** Provides the framework to automatically solve disputes before going into the resolution module. The [BondEscalationModule](https://github.com/defi-wonderland/prophet-modules/blob/dev/solidity/contracts/modules/dispute/BondEscalationModule.sol) manages disputes by allowing users to bond tokens to vote for the user that is telling the truth. The only modification we need to add is that only one propose per request can be solved through escalation. We can safely remove this.

Reviewer: I am not clear if you are suggesting we need or that we do not need more than one.
```suggestion
- **DisputeModule:** Provides the framework to automatically solve disputes before going into the resolution module. The [BondEscalationModule](https://github.com/defi-wonderland/prophet-modules/blob/dev/solidity/contracts/modules/dispute/BondEscalationModule.sol) manages disputes by allowing users to bond tokens to vote for the user that is telling the truth. The only modification we need to add is that only one proposal per request can be solved through escalation. We can safely remove this.
```

> - **ResolutionModule:** Is the last instance to solve disputes and the ultimate source of truth. In our case we will be using the [ArbitratorModule](https://github.com/defi-wonderland/prophet-modules/blob/dev/solidity/contracts/modules/resolution/ArbitratorModule.sol) (Council) with a custom periphery contract so that The Graph's Arbitrator can push resolutions to the few disputes that can't be solved by escalation.

Reviewer: I think this is what is meant. 
```suggestion
- **ResolutionModule:** Used as the last resort to solve disputes and the ultimate source of truth. In our case we will be using the [ArbitratorModule](https://github.com/defi-wonderland/prophet-modules/blob/dev/solidity/contracts/modules/resolution/ArbitratorModule.sol) (Council) with a custom periphery contract so that The Graph's Arbitrator can resolve escaleted disputes.
```

> - **FinalityModule:** Executes code when a request is fully finished. We will be forking the [CallbackModule](https://github.com/defi-wonderland/prophet-modules/blob/dev/solidity/contracts/modules/finality/CallbackModule.sol) to add the events and functions necessary to be able to index it into the subgraph. This new FinalityModule will behave like the previous DataEdge contract. It will receive finalized requests for each chainId-block and emit events for them for the subgraph to index. Gated functions will also be included to allow The Graph's Arbitrator to amend any errors or publish data needed in case an emergency happens.

Reviewer: ```suggestion
- **FinalityModule:** Executes code when a request is fully finished. We will be forking the [CallbackModule](https://github.com/defi-wonderland/prophet-modules/blob/dev/solidity/contracts/modules/finality/CallbackModule.sol) to add the events and functions necessary to be able to index it into the subgraph. This new FinalityModule will behave like the previous DataEdge contract. It will receive finalized requests for each chainId-block and emit events for them for the subgraph to index. Gated functions will also be included to allow The Graph's Arbitrator to amend any errors or publish data needed in case of emergency.
```

- **AccountingExtension:** Provides an interface to bond the user's tokens. We will support bonding in the Horizon Staking smart contract that is being developed by The Graph. We will be using and modifying the [BondEscalationAccounting](https://github.com/defi-wonderland/prophet-modules/blob/dev/solidity/contracts/extensions/BondEscalationAccounting.sol) module for this because it integrates fully with the BondEscalationModule on disputes.

### **Subgraph**

> - **The EBO** subgraph will contain the indexed data generated by the FinalityModule. It will register new data published from prophet and also amendments pushed by the Council.

Reviewer: Slightly strange bolding to me? Either just EBO or none needed?

### **Periphery**

> - **EBORequestCreator:** Will simplify and validate the creation of requests plus also allowing The Graph to add rewards to data providers in the future if they wish to. Checks that the requested epoch hasn't been requested before and that it has already started. The EBORequestCreator will also include the accepted chains. The Graph's Arbitrator will have the ability to add or remove them.

Reviewer: One minor tweak inline. Second question 'they' is 'The Graph'? Normally we identify a particular group and consider 'The Graph' more like an ecosystem than an entity or group? I think this sentance might need some adjustment?
```suggestion
- **EBORequestCreator:** Will simplify and validate the creation of requests and also allow The Graph to add rewards to data providers in the future if they wish to. Checks that the requested epoch hasn't been requested before and that it has already started. The EBORequestCreator will also include the accepted chains. The Graph's Arbitrator will have the ability to add or remove them.
```

> 

### System limitations

The system will not handle requests that escalate to the Arbitrator for final resolution, as this isn’t automated by the off-chain agent and the Arbitrator must resolve the dispute manually. Additionally, once an epoch concludes, the system will not manage any pending requests.

![EBO-high-level.jpeg](../assets/gip-0080/Optimistic%20EBO%20high-level.jpeg)

The development of the EBO Agent will consist of two modules: 

- **DisputeModule:** The DisputeModule will handle all dispute-related activities and will operate in a fully reactive manner based on incoming events. It is important to note that all dispute logic will occur on the protocol chain. Hence, the DisputeModule will interact exclusively with the protocol chain, which in this case is Arbitrum.
- **BlockNumberModule:** The BlockNumberModule will be responsible solely for computing blocks based on timestamps on the blockchains supported by The Graph, including both EVM and non-EVM chains. For EVM chains, RPC calls will be utilized to extract the necessary data. For non-EVM chains, a third-party service.

# **Dependencies**

> The [horizon staking contract](https://github.com/graphprotocol/contracts/pull/944), as described in GIP-0066, needs to be deployed to integrate with it.

Author: @RembrandtK I see that GIP-0066 and GIP-0068 are now merged onto main. Will add the links to the GIPs

# **Validation**

### **Testnet**

> We propose a trial period on testnet for the new Optimistic EBO implementation, with the participation of volunteer indexers running the new instances. The goal is to test as many connected chains to the system as possible.

Reviewer: ```suggestion
We propose a trial period on testnet for the new Optimistic EBO implementation, with the participation of volunteer indexers running the new instances. The goal is to test as many connected chains with the system as possible.
```

> ### **Internal Reviews**

Reviewer: Great GIP, as you can tell my feedback is mainly minor tweaks rather than seeing anything of substance wrong.

I wonder if we want to avoid the 'process and planning' of this section being a part of a GIP, and try to keep the GIP about a technical proposal. Not a big deal in this case, but what if those plans or details change? Do we update the GIP? If so, unnecessary churn, if not, cannot rely on what is written here, and it might mislead.

Author: Thank you @RembrandtK. Agree that it's a good idea to remove this! We are also removing the Testnet phase as it's not mandatory and it's something we will discuss at a later stage

A team of experienced developers from Wonderland, who have not been previously involved in the project, will carry out the following tasks:

- In-depth analysis of EBO requirements from a security perspective.
- Security review of each contract developed for the EBO use case.
- Implementation of necessary fixes based on relevant findings.
- Testing campaign including property-based testing with Echidna and an integration testing suite.

### **Audit**

The Graph Foundation and core devs will audit the code through an external security company. Based on the relevant findings, Wonderland will implement the necessary fixes.

# Who is Wonderland?

We’re a group of developers, researchers, and data scientists with one thing in common: we all love building cool sh*t. [DeFi sucks](https://defi.sucks/), but we are here to make it better.

Our mission is to discover, partner, and empower innovators in the creation of open, permissionless, and decentralized financial solutions. Our pledge is to stand by our partners, supporting them in every way we can.

> We have partnered with some of the most successful and promising protocols in Web3 – including Aztec, Optimism, Everclear, Safe, MakerDAO, Gitcoin, Layer 3, Eigen Layer, Keep3r, Reflexer, and Yearn – to find solutions to complex engineering challenges and help them reach their full potential.

Reviewer: Would be nice to link to them, so easy to look them all up. :)
